zenoh/src/net/routing/dispatcher/pubsub.rs

```diff
@@ -177,61 +180,24 @@ pub(crate) fn undeclare_subscription(
     }
 }
 
-fn compute_data_routes_(tables: &Tables, routes: &mut DataRoutes, expr: &mut RoutingExpr) {
-    let indexes = tables.hat_code.get_data_routes_entries(tables);
-
-    let max_idx = indexes.routers.iter().max().unwrap();
-    routes
-        .routers
-        .resize_with((*max_idx as usize) + 1, || Arc::new(HashMap::new()));
-
-    for idx in indexes.routers {
-        routes.routers[idx as usize] =
-            tables
-                .hat_code
-                .compute_data_route(tables, expr, idx, WhatAmI::Router);
-    }
-
-    let max_idx = indexes.peers.iter().max().unwrap();
-    routes
-        .peers
-        .resize_with((*max_idx as usize) + 1, || Arc::new(HashMap::new()));
-
-    for idx in indexes.peers {
-        routes.peers[idx as usize] =
-            tables
-                .hat_code
-                .compute_data_route(tables, expr, idx, WhatAmI::Peer);
-    }
-
-    let max_idx = indexes.clients.iter().max().unwrap();
-    routes
-        .clients
-        .resize_with((*max_idx as usize) + 1, || Arc::new(HashMap::new()));
-
-    for idx in indexes.clients {
-        routes.clients[idx as usize] =
-            tables
-                .hat_code
-                .compute_data_route(tables, expr, idx, WhatAmI::Client);
-    }
-}
-
 pub(crate) fn compute_data_routes(tables: &Tables, expr: &mut RoutingExpr) -> DataRoutes {
     let mut routes = DataRoutes::default();
-    compute_data_routes_(tables, &mut routes, expr);
+    tables
+        .hat_code
+        .compute_data_routes(tables, &mut routes, expr);
     routes
 }
 
 pub(crate) fn update_data_routes(tables: &Tables, res: &mut Arc<Resource>) {
-    if res.context.is_some() {
+    if res.context.is_some() && !res.expr().contains('*') && res.has_subs() {
```

Comment:
`expr()` returns `&str`, correct?
If `expr()` would return `keyexpr` you could use https://github.com/eclipse-zenoh/zenoh/blob/f90587b580a09b4ea66bb0f796a55699278c49b9/commons/zenoh-keyexpr/src/key_expr/borrowed.rs#L133

and rely on the key-expr logic instead of using `contains()` everywhere. It's more a matter of concern encapsulation than anything else.

Reply:
I fully agree. The problem is that: 
- the resource tree cannot store keyexprs for now without a big refactor. 
- doing a try_into keyexpr each time accessing the expr when not needed has a significant cost.

So until we refactor the resource tree, it's far more efficient to do it like this.